File: 0222.Count Complete Tree Nodes/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if(root == nullptr)
            return 0;
        TreeNode *left = root, *right = root;
        int leftHeight = 0, rightHeight = 0;
        while(left){
            left = left->left;
            leftHeight++;
        }
        while(right){
            right = right->right;
            rightHeight++;
        }
        if(leftHeight == rightHeight)
            return pow(2, leftHeight) - 1;
        else
            return 1 + countNodes(root->left) + countNodes(root->right);
    }
};

File: offer/JZ27 二叉树的镜像/tree.java
using System;
using System.Collections.Generic;

/*
public class TreeNode
{
	public int val;
	public TreeNode left;
	public TreeNode right;

	public TreeNode (int x)
	{
		val = x;
	}
}
*/

class Solution {
    public void reverse(TreeNode root){
        if(root == null)
            return;
        TreeNode left = root.left;
        root.left = root.right;
        root.right = left;
        reverse(root.left);
        reverse(root.right);
    }
    public TreeNode Mirror (TreeNode root) {
        reverse(root);
        return root;
    }
}